alf/trainers/off_policy_trainer.py

```diff
@@ -12,230 +12,116 @@
-                env_f=env_f,
-                algorithm=algorithm,
-                unroll_length=unroll_length,
-                debug_summaries=debug_summaries,
-                summarize_grads_and_vars=summarize_grads_and_vars)
-        replayer = driver.exp_replayer
-
-        checkpointer = tfa_common.Checkpointer(
-            ckpt_dir=os.path.join(train_dir, 'algorithm'),
-            algorithm=algorithm,
-            metrics=metric_utils.MetricsGroup(driver.get_metrics(), 'metrics'),
-            global_step=global_step)
-        checkpointer.initialize_or_restore()
-
-        if not use_tf_functions:
-            tf.config.experimental_run_functions_eagerly(True)
-
-        env.reset()
-        driver.start()
-        time_step = driver.get_initial_time_step()
-        policy_state = driver.get_initial_policy_state()
+class OffPolicyTrainer(Trainer):
+    def __init__(self,
+                 root_dir,
+                 initial_collect_steps=0,
+                 num_updates_per_train_step=4,
+                 unroll_length=8,
+                 mini_batch_length=20,
+                 mini_batch_size=256,
+                 clear_replay_buffer=True):
+        """Abstract base class for off policy trainer
+
+        Args:
+            initial_collect_steps (int): if positive, number of steps each single
+                environment steps before perform first update
+            num_updates_per_train_step (int): number of optimization steps for
+                one iteration
+            unroll_length (int): number of time steps each environment proceeds per
+                iteration. The total number of time steps from all environments per
+                iteration can be computed as: `num_envs` * `env_batch_size`
+                * `unroll_length`.
+            mini_batch_size (int): number of sequences for each minibatch
+            mini_batch_length (int): the length of the sequence for each
+                sample in the minibatch
+            clear_replay_buffer (bool): whether use all data in replay buffer to
+                perform one update and then wiped clean
+
+        """
+        super().__init__(root_dir)
+        self._initial_collect_steps = initial_collect_steps
+        self._num_updates_per_train_step = num_updates_per_train_step
+        self._unroll_length = unroll_length
+        self._mini_batch_length = mini_batch_length
+        self._mini_batch_size = mini_batch_size
+
+        def _get_all_exp():
+            replay_buffer = self._driver.exp_replayer
+            experience = replay_buffer.replay_all()
+            replay_buffer.clear()
+            return experience
+
+        def _sample_exp():
+            replay_buffer = self._driver.exp_replayer
+            return replay_buffer.replay(
+                sample_batch_size=self._mini_batch_size,
+                mini_batch_length=self._mini_batch_length)[0]
+
+        self._get_exp = _get_all_exp if clear_replay_buffer else _sample_exp
```

Comment:
We can let the subclass implement self._get_exp because base class does not use  _get_exp() and different subclass need different _get_exp. And perhaps _get_exp() should just be part of train_iter() to make things simple.

Reply:
it's common part for sub classes